cyral/resource_cyral_integration_idp_saml_draft.go

```diff
@@ -0,0 +1,281 @@
+package cyral
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+
+	"github.com/cyralinc/terraform-provider-cyral/client"
+)
+
+// Note that although the IdP type is SAML as seen from the control plane UI, we
+// use GenericSAML in many variables, to make it more specific.
+//
+// The SAML IdP type is different from all others, in that it is _generic_. It
+// allows the user to configure any type of IdP integration, as long as they
+// provide the IdP metadata.
+//
+// This resource only covers the _draft_ of the integration. The draft is
+// created on the control plane, and the SP metadata is generated. To complete
+// the draft, one must perform the following steps:
+//
+// 1. Provide the SP metadata generated by this resource to the IdP.
+//
+// 2. Provide the IdP metadata to the `cyral_integration_idp_saml` resource.
+//
+
+type CreateGenericSAMLDraftRequest struct {
+	DisplayName              string                  `json:"displayName"`
+	DisableIdPInitiatedLogin bool                    `json:"disableIdPInitiatedLogin"`
+	IdpType                  string                  `json:"idpType,omitempty"`
+	Attributes               *RequiredUserAttributes `json:"attributes,omitempty"`
+}
+
+func (req *CreateGenericSAMLDraftRequest) ReadFromSchema(d *schema.ResourceData) error {
+	req.DisplayName = d.Get("display_name").(string)
+	req.DisableIdPInitiatedLogin = d.Get("disable_idp_initiated_login").(bool)
+	req.IdpType = d.Get("idp_type").(string)
+
+	attributes, err := RequiredUserAttributesFromSchema(d)
+	if err != nil {
+		return err
+	}
+	req.Attributes = attributes
+
+	return nil
+}
+
+type GenericSAMLDraftResponse struct {
+	Draft GenericSAMLDraft `json:"draft"`
+}
+
+func (resp *GenericSAMLDraftResponse) WriteToSchema(d *schema.ResourceData) error {
+	d.SetId(resp.Draft.ID)
+	if err := d.Set("display_name", resp.Draft.DisplayName); err != nil {
+		return err
+	}
+	if err := d.Set("disable_idp_initiated_login", resp.Draft.DisableIdPInitiatedLogin); err != nil {
+		return err
+	}
+	if err := d.Set("sp_metadata", resp.Draft.SPMetadata.XMLDocument); err != nil {
+		return err
+	}
+	if err := d.Set("idp_type", resp.Draft.IdpType); err != nil {
+		return err
+	}
+	if resp.Draft.Attributes != nil && len(d.Get("attributes").(*schema.Set).List()) > 0 {
```

Comment:
It's because otherwise the response will contain the default values, and then when we write to the schema, and the user runs `terraform refresh` or `terraform apply`, the plan will not be empty. To avoid that, we can either do the above, or require that the user specify the `attributes` set, as below:

```terraform
resource "cyral_integration_idp_saml_draft" "myresource" {
    ...
    attributes{}
}
```

I have added an ACC test that captures this case, to prevent regressions.